File: dotnet/src/Envdoctor/Models/Origin.cs
namespace Envdoctor.Models;

public enum OriginKind
{
    Definition,
    Reference,
    Usage,
}

public static class OriginKindExtensions
{
    public static string AsStr(this OriginKind k) => k switch
    {
        OriginKind.Definition => "definition",
        OriginKind.Reference => "reference",
        OriginKind.Usage => "usage",
        _ => "usage",
    };
}

public enum OriginFormat
{
    Dotenv,
    DockerCompose,
    GithubActions,
    Kubernetes,
    Source,
}

public static class OriginFormatExtensions
{
    public static string AsStr(this OriginFormat f) => f switch
    {
        OriginFormat.Dotenv => "dotenv",
        OriginFormat.DockerCompose => "docker-compose",
        OriginFormat.GithubActions => "github-actions",
        OriginFormat.Kubernetes => "kubernetes",
        OriginFormat.Source => "source",
        _ => "source",
    };
}

/// Where a variable name was seen, and in what role. Values are intentionally
/// NOT carried here.
public sealed class Origin
{
    public required string FilePath { get; set; }
    public int? Line { get; set; }
    public OriginKind Kind { get; set; }
    public string? Environment { get; set; }
    public OriginFormat? Format { get; set; }
    public string? Subkind { get; set; }

    public Origin Clone() => (Origin)MemberwiseClone();

    public static Origin NewDefinition(string filePath, int? line, string? environment) =>
        new()
        {
            FilePath = filePath,
            Line = line,
            Kind = OriginKind.Definition,
            Environment = environment,
            Format = OriginFormat.Dotenv,
        };

    public static Origin NewReference(string filePath, int? line, string? environment) =>
        new()
        {
            FilePath = filePath,
            Line = line,
            Kind = OriginKind.Reference,
            Environment = environment,
            Format = OriginFormat.Dotenv,
        };

    public static Origin NewUsage(string filePath, int? line, OriginFormat format) =>
        new()
        {
            FilePath = filePath,
            Line = line,
            Kind = OriginKind.Usage,
            Format = format,
        };

    public Origin WithSubkind(string subkind)
    {
        Subkind = subkind;
        return this;
    }
}

File: swift/Sources/EnvdoctorKit/Models/Origin.swift
import Foundation

/// Where a variable name was seen, and in what role.
public struct Origin: Equatable {
    public enum Kind: String, Equatable {
        case definition, reference, usage
    }

    public enum Format: String, Equatable {
        case dotenv, dockerCompose = "docker-compose", githubActions = "github-actions", kubernetes, source
    }

    /// Path to the file where the variable appeared (repo-relative when possible).
    public var filePath: String
    /// 1-based line number when known.
    public var line: Int?
    public var kind: Kind
    /// Environment label (dotenv files only, e.g. "development", "production").
    public var environment: String?
    /// The format the origin came from.
    public var format: Format?
    /// Format-specific detail (e.g. "secrets" vs "vars" for GitHub Actions).
    public var subkind: String?

    public init(filePath: String, line: Int? = nil, kind: Kind, environment: String? = nil, format: Format? = nil, subkind: String? = nil) {
        self.filePath = filePath
        self.line = line
        self.kind = kind
        self.environment = environment
        self.format = format
        self.subkind = subkind
    }
}
